globus_automate_client/models.py

```diff
@@ -0,0 +1,387 @@
+    TimestampLessThanEquals: t.Optional[str]
+    TimestampLessThanEqualsPath: t.Optional[str]
+    TimestampGreaterThanEquals: t.Optional[str]
+    TimestampGreaterThanEqualsPath: t.Optional[str]
+    IsNull: t.Optional[str]
+    IsPresent: t.Optional[str]
+    IsNumeric: t.Optional[str]
+    IsString: t.Optional[str]
+    IsBoolean: t.Optional[str]
+    IsTimestamp: t.Optional[str]
+
+    class Config:
+        extra = Extra.forbid
+
+    _enforce_jsonpath = validator(
+        "Variable",
+        "StringEqualsPath",
+        "StringLessThanPath",
+        "StringGreaterThanPath",
+        "StringLessThanEqualsPath",
+        "StringGreaterThanEqualsPath",
+        "NumericEqualsPath",
+        "NumericLessThanPath",
+        "NumericGreaterThanPath",
+        "NumericLessThanEqualsPath",
+        "NumericGreaterThanEqualsPath",
+        "BooleanEqualsPath",
+        "TimestampEqualsPath",
+        "TimestampLessThanPath",
+        "TimestampGreaterThanPath",
+        "TimestampLessThanEqualsPath",
+        "TimestampGreaterThanEqualsPath",
+        allow_reuse=True,
+    )(enforce_jsonpath)
+
+    @root_validator(skip_on_failure=True)
+    def validate_only_one_rule_is_set(cls, values):
+        set_fields = {
+            k
+            for k, v in values.items()
+            if k not in {"Variable", "Next"} and v is not None
+        }
+        if len(set_fields) != 1:
+            raise ValueError(
+                f'A "ChoiceRule" may only contain one expression. Found {", ".join(set_fields)}'
+            )
+        return values
+
+
+class DataTestExpressionChoiceRule(ChoiceRule):
+    Next: str
+
+
+class BooleanExpressionChoiceRule(BaseModel):
+    """
+    A Boolean Expression is a JSON object which contains a field named "And",
+    "Or", or "Not". If the field name is "And" or "Or", the value MUST be an
+    non-empty object array of Choice Rules, which MUST NOT contain "Next"
+    fields; the interpreter processes the array elements in order, performing
+    the boolean evaluations in the expected fashion, ceasing array processing
+    when the boolean value has been unambiguously determined.
+
+    The value of a Boolean Expression containing a "Not" field MUST be a single
+    Choice Rule, that MUST NOT contain a "Next" field; it returns the inverse
+    of the boolean to which the Choice Rule evaluates.
+    """
+
+    And: t.Optional[t.List[ChoiceRule]] = Field(None, min_items=1)
+    Or: t.Optional[t.List[ChoiceRule]] = Field(None, min_items=1)
+    Not: t.Optional[ChoiceRule] = None
+    Next: str
+
+    @root_validator(skip_on_failure=True)
+    def validate_only_one_boolean_choice_rule_set(cls, values):
+        set_fields = {
+            k for k, v in values.items() if k not in {"Next"} and v is not None
+        }
+        if len(set_fields) == 0:
+            raise ValueError(
+                'One of "And", "Or", or "Not" must be specified in a BooleanExpression'
+            )
```

Comment:
I've updated to handle the first case -- now the check if validating that exactly 1 of the non-Next fields is set to a non-None value.

I'm not sure of how to handle number 2. It's a built in assumption that if the model defines a default value and the input data does not explicitly provide that value, the resulting model instance will contain that default value. If we go ahead and re-serialize a FlowDefinition model it will contain `None` values for all optional/exclusive fields that were unset. There is an option to re-serialize the model by omittitng unset values if there's interest in only getting back fields that were explicitly set. I updated tests with that usage to make sure that the dict that goes in is what comes back out.

Reply:
I've seen the `exclude_unset` arg and I'm anticipating that will always be used. Is there a way to specify an alternate sentinel value besides `None`? I think `Ellipsis` has special meaning to pydantic, but perhaps it, or an empty tuple, might be a usable sentinel that can't be created by JSON or YAML inputs?